pageserver/src/tenant.rs

```diff
@@ -350,9 +361,604 @@ impl Drop for TimelineUninitMark {
+                    .unwrap_or_default()
+                    .parse::<TimelineId>()
+                    .with_context(|| {
+                        format!(
+                            "Could not parse timeline id out of the timeline dir name {}",
+                            timeline_dir.display()
+                        )
+                    })?;
+                let timeline_uninit_mark_file = self
+                    .conf
+                    .timeline_uninit_mark_file_path(self.tenant_id, timeline_id);
+                if timeline_uninit_mark_file.exists() {
+                    info!(
+                        "Found an uninit mark file for timeline {}/{}, removing the timeline and its uninit mark",
+                        self.tenant_id, timeline_id
+                    );
+                    if let Err(e) =
+                        remove_timeline_and_uninit_mark(&timeline_dir, &timeline_uninit_mark_file)
+                    {
+                        error!("Failed to clean up uninit marked timeline: {e:?}");
+                    }
+                    continue;
+                }
+
+                // TODO There was also remove if dir is empty logic
+                // TODO if attach failed we can be missing metadata here
+
+                let file_name = entry.file_name();
+                if let Ok(timeline_id) =
+                    file_name.to_str().unwrap_or_default().parse::<TimelineId>()
+                {
+                    let metadata = load_metadata(self.conf, timeline_id, self.tenant_id)
+                        .context("failed to load metadata")?;
+                    timelines_to_load.insert(timeline_id, metadata);
+                } else {
+                    // A file or directory that doesn't look like a timeline ID
+                    warn!(
+                        "unexpected file or directory in timelines directory: {}",
+                        file_name.to_string_lossy()
+                    );
+                }
+            }
+        }
+
+        // Sort the array of timeline IDs into tree-order, so that parent comes before
+        // all its children.
+        let sorted_timelines = tree_sort_timelines(timelines_to_load)?;
+        // FIXME if we init some of the timelines and one of them is broken we
+        //      potentially will end up with partially launched wal receivers.
+        //    1. Either move it out of initialize_with_lock
+        //    2. Or pass it as a flag
+        // FIXME original collect_timeline_files contained one more check:
+        //    1. "Timeline has no ancestor and no layer files"
+
+        // XXX get rid of enable_background_jobs
+        let enable_background_jobs = !sorted_timelines.is_empty();
+
+        for (timeline_id, local_metadata) in sorted_timelines {
+            // FIXME should we fail load of whole tenant if one timeline failed?
+            //   consider branch hierarchy. Maybe set one to broken and others to Paused or something
```

Comment:
This loop rougly corresponds to current `main`'s `collect_timelines_for_tenant`.
There, we log & skip the timeline, i.e., it continues to reside on disk without an in-memory representation.


https://github.com/neondatabase/neon/blob/7b7f84f1b4643f0417b6ccc1793cc0c38aeb55db/pageserver/src/tenant_mgr.rs#L742-L766

In my opinion, that's a lurking consistency problem, and erroring out is a clear improvement.

Suggestion: let's remove this FIXME. The worst cacse is that the tenant is `Broken`, so, blast radius is constrainted to one tenant.
We can revisit this in the future.

Reply:
+1 Lets error out instead of skipping